### PR TITLE
Weak-Link the Immediate Re-Exports of a Weak-Linked Module

### PR DESCRIPTION
**Pull request by CodaFi**
When a module is imported @_weakLinked, its re-exported peer modules have their definitions imported with strong linkage which can often defeat the point of weak linking the parent in the first place. Allow the weak-linkage to float to the immediate re-exports of the @_weakLinked module.

Resolves rdar://117166194

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -3441,6 +3441,17 @@ bool SourceFile::importsModuleAsWeakLinked(const ModuleDecl *module) const {
...
     if (module == clangModule)
       return true;
+
+    // Traverse the exported modules of this weakly-linked module to ensure
+    // that we weak-link declarations from its exported peers.
+    SmallVector<ImportedModule, 8> reexportedModules;
+    importedModule->getImportedModules(reexportedModules,
+                                       ModuleDecl::ImportFilterKind::Exported);
+    for (const ImportedModule &reexportedModule : reexportedModules) {
+      if (!module->isNonSwiftModule() &&
```
Comment (tshortli) [inline lib/AST/Module.cpp:3451]: Why do we need this? Weak linking is expected to apply to clang modules as well.

**Review by tshortli — COMMENTED** [review]
Thanks for working on this!

File: lib/AST/Module.cpp (hunk 1 above)
Comment (CodaFi) [reply lib/AST/Module.cpp:3451]: I wanted to restrict the behavior change in this patch, but I can remove this check and add a test

**Review by tshortli — APPROVED** [review]

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**